File: LC/python/Longest Valid Parentheses.py
class Solution:
    # @param s, a string
    # @return an integer
    def longestValidParentheses(self, s):
        stack = []
        tmp = -1
        maxlen = 0
        for i in range(0,len(s)):
            if s[i] == '(':
                stack.append(i)
            else:
                if stack == []:
                    tmp = i
                else:
                    stack.pop()
                    if stack ==[]:
                        maxlen = max(maxlen, i-tmp)
                    else:
                        maxlen = max(maxlen, i-stack[len(stack)-1])
        return maxlen
                    

File: LC/java/Longest Valid Parentheses.java
public class Solution {
    public int longestValidParentheses(String s) {
        int tmp  = -1, maxlen = 0;
        Stack<Integer> q = new Stack<Integer>();
        
        for(int i = 0; i < s.length(); i++){
            if(s.charAt(i) == '('){
                q.add(i);
            }
            else{
                if(q.size()==0){
                    tmp = i;
                }
                else{
                    q.pop();
                    if(q.size()==0){
                        maxlen = Math.max(maxlen, i-tmp);
                    }
                    else{
                        maxlen = Math.max(maxlen, i-q.get(q.size()-1));
                    }
                }
            }
        }
        return maxlen;
    }
}
